modules/generator/registry/native_histogram.go

```diff
@@ -0,0 +1,336 @@
+		traceIDLabelName = "traceID"
+	}
+
+	return &nativeHistogram{
+		metricName:       name,
+		series:           make(map[uint64]*nativeHistogramSeries),
+		onAddSerie:       onAddSeries,
+		onRemoveSerie:    onRemoveSeries,
+		traceIDLabelName: traceIDLabelName,
+		buckets:          buckets,
+	}
+}
+
+func (h *nativeHistogram) ObserveWithExemplar(labelValueCombo *LabelValueCombo, value float64, traceID string, multiplier float64) {
+	hash := labelValueCombo.getHash()
+
+	h.seriesMtx.Lock()
+	defer h.seriesMtx.Unlock()
+
+	s, ok := h.series[hash]
+	if ok {
+		h.updateSeries(s, value, traceID, multiplier)
+		return
+	}
+
+	if !h.onAddSerie(h.activeSeriesPerHistogramSerie()) {
+		return
+	}
+
+	newSeries := h.newSeries(labelValueCombo, value, traceID, multiplier)
+	h.series[hash] = newSeries
+}
+
+func (h *nativeHistogram) newSeries(labelValueCombo *LabelValueCombo, value float64, traceID string, multiplier float64) *nativeHistogramSeries {
+	newSeries := &nativeHistogramSeries{
+		// TODO move these labels in HistogramOpts.ConstLabels?
+		labels: labelValueCombo.getLabelPair(),
+		promHistogram: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    h.name(),
+			Help:    "Native histogram for metric " + h.name(),
+			Buckets: h.buckets,
+			// TODO check if these values are sensible and break them out
+			NativeHistogramBucketFactor:     1.1,
+			NativeHistogramMaxBucketNumber:  100,
+			NativeHistogramMinResetDuration: 15 * time.Minute,
+		}),
+		lastUpdated: atomic.NewInt64(0),
+	}
+
+	h.updateSeries(newSeries, value, traceID, multiplier)
+
+	return newSeries
+}
+
+func (h *nativeHistogram) updateSeries(s *nativeHistogramSeries, value float64, traceID string, multiplier float64) {
+	for i := 0.0; i < multiplier; i++ {
+		s.promHistogram.(prometheus.ExemplarObserver).ObserveWithExemplar(
+			value,
+			map[string]string{h.traceIDLabelName: traceID},
+		)
+	}
+	s.lastUpdated.Store(time.Now().UnixMilli())
+}
+
+func (h *nativeHistogram) name() string {
+	return h.metricName
+}
+
+func (h *nativeHistogram) collectMetrics(appender storage.Appender, timeMs int64, externalLabels map[string]string) (activeSeries int, err error) {
+	h.seriesMtx.Lock()
+	defer h.seriesMtx.Unlock()
+
+	labelsCount := 0
+	if h.series[0] != nil {
+		labelsCount = len(h.series[0].labels.names)
+	}
+	lbls := make(labels.Labels, 1+len(externalLabels)+labelsCount)
+	lb := labels.NewBuilder(lbls)
+	activeSeries = 0
+
+	lb.Set(labels.MetricName, h.metricName)
+
+	// set external labels
+	for name, value := range externalLabels {
+		lb.Set(name, value)
+	}
+
+	for _, s := range h.series {
+
+		// Set series-specific labels
+		for i, name := range s.labels.names {
+			lb.Set(name, s.labels.values[i])
+		}
+
+		// Extract histogram
+		encodedMetric := &dto.Metric{}
+
+		// Encode to protobuf representation
+		err = s.promHistogram.Write(encodedMetric)
+		if err != nil {
+			return activeSeries, err
+		}
+		s.histogram = encodedMetric.GetHistogram()
```

Comment:
is there a reason to store this as a field on the struct? it seems like we only use it in this method. if that's true we should remove it from the struct

Reply:
This was added so that we could remember which histograms were sent.  The `Reset` call below captures this, and storing the state on the struct was a way for us to get around this issue.  If we don't know which buckets have had their exemplars reported, they get updated with the most recent timestamp and sent again, since the series map contains keeps the exemplars.

Reply:
yeah, i'm not seeing this. it seems only referenced in this method which makes me think we should just use a local var. perhaps i'm missing something. we can sync on this.

Reply:
Yeah, let's sync.